File: CS/CustomChartDataSourceWin/CustomChartDataSource.cs
using System;
using System.Text;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraPivotGrid;
using System.Collections.Generic;

namespace DXSample {
    public class CustomChartDataSource :IBindingList, ITypedList, IDisposable {
        private PivotGridControl fPivot;
        private List<object[]> fData;

        public CustomChartDataSource(PivotGridControl source) {
            fPivot = source;
            CreateChartData();
            fPivot.CellSelectionChanged += OnPivotCellSelectionChanged;
        }

        void OnPivotCellSelectionChanged(object sender, EventArgs e) {
            CreateChartData();
            RaiseListChanged();
        }

        void CreateChartData() {
            fData = fPivot.OptionsSelection.MultiSelect && fPivot.OptionsChartDataSource.SelectionOnly ? CreateChartDataFromMultiselection() :
                CreateChartDataFromSelection();
        }

        List<object[]> CreateChartDataFromSelection() {
            List<object[]> result = new List<object[]>();
            Rectangle selection = fPivot.OptionsChartDataSource.SelectionOnly ? fPivot.Cells.Selection :
                new Rectangle(0, 0, fPivot.Cells.ColumnCount, fPivot.Cells.RowCount);
            for (int col = selection.Left; col < selection.Right; col++)
                for (int row = selection.Top; row < selection.Bottom; row++) {
                    object[] data = GetDataFromCell(fPivot.Cells.GetCellInfo(col, row));
                    if (data != null)
                        result.Add(data);
                }
            return result;
        }

        List<object[]> CreateChartDataFromMultiselection() {
            List<object[]> result = new List<object[]>();
            foreach (Point cell in fPivot.Cells.MultiSelection.SelectedCells) {
                object[] data = GetDataFromCell(fPivot.Cells.GetCellInfo(cell.X, cell.Y));
                if (data != null)
                    result.Add(data);
            }
            return result;
        }

        object[] GetDataFromCell(PivotCellEventArgs cell) {
            if (SkipCellValue(cell)) return null;
            object[] result = new object[3];
            result[0] = string.Concat(fPivot.OptionsChartDataSource.ChartDataVertical ? GetFieldValue(cell, false) : GetFieldValue(cell, true),
                fPivot.GetFieldsByArea(PivotArea.DataArea).Count > 1 ? string.Concat(" | ", cell.DataField.Caption) : string.Empty);
            result[1] = fPivot.OptionsChartDataSource.ChartDataVertical ? GetFieldValue(cell, true) : GetFieldValue(cell, false);
            result[2] = Convert.ToDecimal(cell.Value);
            return result;
        }

        private static string GetFieldValue(PivotCellEventArgs cell, bool column) {
            PivotGridField[] fields = column ? cell.GetColumnFields() : cell.GetRowFields();
            StringBuilder result = new StringBuilder();
            foreach (PivotGridField field in fields)
                result.AppendFormat("{0} | ", cell.GetFieldValue(field));
            if (result.Length > 3)
                result.Remove(result.Length - 3, 3);
            if (column)
                switch (cell.ColumnValueType) {
                    case PivotGridValueType.CustomTotal:
                        return result.AppendFormat(" Total ({0})", cell.ColumnCustomTotal.SummaryType).ToString();
                    case PivotGridValueType.GrandTotal:
                        return "Grand Total";
                    case PivotGridValueType.Total:
                        return result.Append(" Total").ToString();
                } else
                switch (cell.RowValueType) {
                    case PivotGridValueType.CustomTotal:
                        return result.AppendFormat(" Total ({0})", cell.RowCustomTotal.SummaryType).ToString();
                    case PivotGridValueType.GrandTotal:
                        return "Grand Total";
                    case PivotGridValueType.Total:
                        return result.Append(" Total").ToString();
                }
            return result.ToString();
        }

        private bool SkipCellValue(PivotCellEventArgs cell) {
            return (!fPivot.OptionsChartDataSource.ShowColumnCustomTotals && cell.ColumnValueType == PivotGridValueType.CustomTotal) ||
                (!fPivot.OptionsChartDataSource.ShowColumnGrandTotals && cell.ColumnValueType == PivotGridValueType.GrandTotal) ||
                (!fPivot.OptionsChartDataSource.ShowColumnTotals && cell.ColumnValueType == PivotGridValueType.Total) ||
                (!fPivot.OptionsChartDataSource.ShowRowCustomTotals && cell.RowValueType == PivotGridValueType.CustomTotal) ||
                (!fPivot.OptionsChartDataSource.ShowRowGrandTotals && cell.RowValueType == PivotGridValueType.GrandTotal) ||
                (!fPivot.OptionsChartDataSource.ShowRowTotals && cell.RowValueType == PivotGridValueType.Total);
        }

        #region IBindingList Members

        void IBindingList.AddIndex(PropertyDescriptor property) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        object IBindingList.AddNew() {
            throw new System.Exception("The method or operation is not implemented.");
        }

        bool IBindingList.AllowEdit {
            get { return false; }
        }

        bool IBindingList.AllowNew {
            get { return false; }
        }

        bool IBindingList.AllowRemove {
            get { return false; }
        }

        void IBindingList.ApplySort(PropertyDescriptor property, ListSortDirection direction) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        int IBindingList.Find(PropertyDescriptor property, object key) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        bool IBindingList.IsSorted {
            get { throw new System.Exception("The method or operation is not implemented."); }
        }

        private ListChangedEventHandler fListChanged;
        protected void RaiseListChanged() {
            if (fListChanged != null)
                fListChanged(this, new ListChangedEventArgs(ListChangedType.Reset, -1));
        }
        event ListChangedEventHandler IBindingList.ListChanged {
            add { fListChanged += value; }
            remove { fListChanged -= value; }
        }

        void IBindingList.RemoveIndex(PropertyDescriptor property) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        void IBindingList.RemoveSort() {
            throw new System.Exception("The method or operation is not implemented.");
        }

        ListSortDirection IBindingList.SortDirection {
            get { throw new System.Exception("The method or operation is not implemented."); }
        }

        PropertyDescriptor IBindingList.SortProperty {
            get { throw new System.Exception("The method or operation is not implemented."); }
        }

        bool IBindingList.SupportsChangeNotification {
            get { return true; }
        }

        bool IBindingList.SupportsSearching {
            get { return false; }
        }

        bool IBindingList.SupportsSorting {
            get { return false; }
        }

        #endregion

        #region IList Members

        int IList.Add(object value) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        void IList.Clear() {
            throw new System.Exception("The method or operation is not implemented.");
        }

        bool IList.Contains(object value) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        int IList.IndexOf(object value) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        void IList.Insert(int index, object value) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        bool IList.IsFixedSize {
            get { return true; }
        }

        bool IList.IsReadOnly {
            get { return true; }
        }

        void IList.Remove(object value) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        void IList.RemoveAt(int index) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        object IList.this[int index] {
            get { return fData[index]; }
            set { throw new System.Exception("The method or operation is not implemented."); }
        }

        #endregion

        #region ICollection Members

        void ICollection.CopyTo(Array array, int index) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        int ICollection.Count {
            get { return fData.Count; }
        }

        bool ICollection.IsSynchronized {
            get { throw new System.Exception("The method or operation is not implemented."); }
        }

        object ICollection.SyncRoot {
            get { throw new System.Exception("The method or operation is not implemented."); }
        }

        #endregion

        #region IEnumerable Members

        IEnumerator IEnumerable.GetEnumerator() {
            return new CustomChartDataSourceEnumerator(fData);
        }

        #endregion

        #region ITypedList Members

        PropertyDescriptorCollection ITypedList.GetItemProperties(PropertyDescriptor[] listAccessors) {
            List<CustomChartDataSourcePropertyDescriptor> properties = new List<CustomChartDataSourcePropertyDescriptor>();
            properties.Add(new CustomChartDataSourcePropertyDescriptor(CustomChartDataSourceProperty.Series));
            properties.Add(new CustomChartDataSourcePropertyDescriptor(CustomChartDataSourceProperty.Arguments));
            properties.Add(new CustomChartDataSourcePropertyDescriptor(CustomChartDataSourceProperty.Values));
            return new PropertyDescriptorCollection(properties.ToArray());
        }

        string ITypedList.GetListName(PropertyDescriptor[] listAccessors) {
            throw new System.Exception("The method or operation is not implemented.");
        }

        #endregion

        #region IDisposable Members

        public void Dispose() {
            if (fPivot != null) {
                fPivot.CellSelectionChanged -= OnPivotCellSelectionChanged;
                fPivot = null;
            }
        }

        #endregion

        public class CustomChartDataSourceEnumerator :IEnumerator {
            List<object[]> fData;
            int fPosition = -1;

            internal CustomChartDataSourceEnumerator(List<object[]> data) {
                fData = data;
            }

            #region IEnumerator Members

            object IEnumerator.Current {
                get {
                    if (fPosition < 0 || fPosition >= fData.Count) return null;
                    return fData[fPosition];
                }
            }

            bool IEnumerator.MoveNext() {
                if (fPosition >= fData.Count) return false;
                fPosition++;
                return true;
            }

            void IEnumerator.Reset() {
                fPosition = -1;
            }

            #endregion
        }

        public class CustomChartDataSourcePropertyDescriptor :PropertyDescriptor {
            private CustomChartDataSourceProperty fPropertyType;

            public CustomChartDataSourcePropertyDescriptor(CustomChartDataSourceProperty propertyType)
                : base(propertyType.ToString(), null) {
                fPropertyType = propertyType;
            }

            public override bool CanResetValue(object component) {
                return false;
            }

            public override Type ComponentType {
                get { return typeof(object[]); }
            }

            public override object GetValue(object component) {
                object[] source = component as object[];
                if (source == null || source.Length != 3) throw new ArgumentException("The array of three values expected");
                return source[(int)fPropertyType];
            }

            public override bool IsReadOnly {
                get { return true; }
            }

            public override Type PropertyType {
                get { return fPropertyType == CustomChartDataSourceProperty.Values ? typeof(decimal) : typeof(string); }
            }

            public override void ResetValue(object component) {
                throw new Exception("The method or operation is not implemented.");
            }

            public override void SetValue(object component, object value) {
                throw new Exception("The method or operation is not implemented.");
            }

            public override bool ShouldSerializeValue(object component) {
                throw new Exception("The method or operation is not implemented.");
            }
        }

        public enum CustomChartDataSourceProperty { Series, Arguments, Values };
}
}

File: VB/CustomChartDataSourceWin/CustomChartDataSource.vb
Imports System
Imports System.Text
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.XtraPivotGrid
Imports System.Collections.Generic

Namespace DXSample

    Public Class CustomChartDataSource
        Implements IBindingList, ITypedList, IDisposable

        Private fPivot As PivotGridControl

        Private fData As List(Of Object())

        Public Sub New(ByVal source As PivotGridControl)
            fPivot = source
            CreateChartData()
            AddHandler fPivot.CellSelectionChanged, AddressOf OnPivotCellSelectionChanged
        End Sub

        Private Sub OnPivotCellSelectionChanged(ByVal sender As Object, ByVal e As EventArgs)
            CreateChartData()
            RaiseListChanged()
        End Sub

        Private Sub CreateChartData()
            fData = If(fPivot.OptionsSelection.MultiSelect AndAlso fPivot.OptionsChartDataSource.SelectionOnly, CreateChartDataFromMultiselection(), CreateChartDataFromSelection())
        End Sub

        Private Function CreateChartDataFromSelection() As List(Of Object())
            Dim result As List(Of Object()) = New List(Of Object())()
            Dim selection As Rectangle = If(fPivot.OptionsChartDataSource.SelectionOnly, fPivot.Cells.Selection, New Rectangle(0, 0, fPivot.Cells.ColumnCount, fPivot.Cells.RowCount))
            For col As Integer = selection.Left To selection.Right - 1
                For row As Integer = selection.Top To selection.Bottom - 1
                    Dim data As Object() = GetDataFromCell(fPivot.Cells.GetCellInfo(col, row))
                    If data IsNot Nothing Then result.Add(data)
                Next
            Next

            Return result
        End Function

        Private Function CreateChartDataFromMultiselection() As List(Of Object())
            Dim result As List(Of Object()) = New List(Of Object())()
            For Each cell As Point In fPivot.Cells.MultiSelection.SelectedCells
                Dim data As Object() = GetDataFromCell(fPivot.Cells.GetCellInfo(cell.X, cell.Y))
                If data IsNot Nothing Then result.Add(data)
            Next

            Return result
        End Function

        Private Function GetDataFromCell(ByVal cell As PivotCellEventArgs) As Object()
            If SkipCellValue(cell) Then Return Nothing
            Dim result As Object() = New Object(2) {}
            result(0) = String.Concat(If(fPivot.OptionsChartDataSource.ProvideDataByColumns, GetFieldValue(cell, False), GetFieldValue(cell, True)), If(fPivot.GetFieldsByArea(PivotArea.DataArea).Count > 1, String.Concat(" | ", cell.DataField.Caption), String.Empty))
            result(1) = If(fPivot.OptionsChartDataSource.ProvideDataByColumns, GetFieldValue(cell, True), GetFieldValue(cell, False))
            result(2) = Convert.ToDecimal(cell.Value)
            Return result
        End Function

        Private Shared Function GetFieldValue(ByVal cell As PivotCellEventArgs, ByVal column As Boolean) As String
            Dim fields As PivotGridField() = If(column, cell.GetColumnFields(), cell.GetRowFields())
            Dim result As StringBuilder = New StringBuilder()
            For Each field As PivotGridField In fields
                result.AppendFormat("{0} | ", cell.GetFieldValue(field))
            Next

            If result.Length > 3 Then result.Remove(result.Length - 3, 3)
            If column Then
                Select Case cell.ColumnValueType
                    Case PivotGridValueType.CustomTotal
                        Return result.AppendFormat(" Total ({0})", cell.ColumnCustomTotal.SummaryType).ToString()
                    Case PivotGridValueType.GrandTotal
                        Return "Grand Total"
                    Case PivotGridValueType.Total
                        Return result.Append(" Total").ToString()
                End Select
            Else
                Select Case cell.RowValueType
                    Case PivotGridValueType.CustomTotal
                        Return result.AppendFormat(" Total ({0})", cell.RowCustomTotal.SummaryType).ToString()
                    Case PivotGridValueType.GrandTotal
                        Return "Grand Total"
                    Case PivotGridValueType.Total
                        Return result.Append(" Total").ToString()
                End Select
            End If

            Return result.ToString()
        End Function

        Private Function SkipCellValue(ByVal cell As PivotCellEventArgs) As Boolean
            Return Not fPivot.OptionsChartDataSource.ProvideColumnCustomTotals AndAlso cell.ColumnValueType = PivotGridValueType.CustomTotal OrElse Not fPivot.OptionsChartDataSource.ProvideColumnGrandTotals AndAlso cell.ColumnValueType = PivotGridValueType.GrandTotal OrElse Not fPivot.OptionsChartDataSource.ProvideColumnTotals AndAlso cell.ColumnValueType = PivotGridValueType.Total OrElse Not fPivot.OptionsChartDataSource.ProvideRowCustomTotals AndAlso cell.RowValueType = PivotGridValueType.CustomTotal OrElse Not fPivot.OptionsChartDataSource.ProvideRowGrandTotals AndAlso cell.RowValueType = PivotGridValueType.GrandTotal OrElse Not fPivot.OptionsChartDataSource.ProvideRowTotals AndAlso cell.RowValueType = PivotGridValueType.Total
        End Function

#Region "IBindingList Members"
        Private Sub AddIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.AddIndex
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private Function AddNew() As Object Implements IBindingList.AddNew
            Throw New Exception("The method or operation is not implemented.")
        End Function

        Private ReadOnly Property AllowEdit As Boolean Implements IBindingList.AllowEdit
            Get
                Return False
            End Get
        End Property

        Private ReadOnly Property AllowNew As Boolean Implements IBindingList.AllowNew
            Get
                Return False
            End Get
        End Property

        Private ReadOnly Property AllowRemove As Boolean Implements IBindingList.AllowRemove
            Get
                Return False
            End Get
        End Property

        Private Sub ApplySort(ByVal [property] As PropertyDescriptor, ByVal direction As ListSortDirection) Implements IBindingList.ApplySort
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private Function Find(ByVal [property] As PropertyDescriptor, ByVal key As Object) As Integer Implements IBindingList.Find
            Throw New Exception("The method or operation is not implemented.")
        End Function

        Private ReadOnly Property IsSorted As Boolean Implements IBindingList.IsSorted
            Get
                Throw New Exception("The method or operation is not implemented.")
            End Get
        End Property

        Private fListChanged As ListChangedEventHandler

        Protected Sub RaiseListChanged()
            If fListChanged IsNot Nothing Then fListChanged(Me, New ListChangedEventArgs(ListChangedType.Reset, -1))
        End Sub

        Private Custom Event ListChanged As ListChangedEventHandler Implements IBindingList.ListChanged
            AddHandler(ByVal value As ListChangedEventHandler)
                fListChanged = [Delegate].Combine(fListChanged, value)
            End AddHandler

            RemoveHandler(ByVal value As ListChangedEventHandler)
                fListChanged = [Delegate].Remove(fListChanged, value)
            End RemoveHandler

            RaiseEvent(ByVal sender As Object, ByVal e As ListChangedEventArgs)
                fListChanged ?(sender, e)
            End RaiseEvent
        End Event

        Private Sub RemoveIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.RemoveIndex
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private Sub RemoveSort() Implements IBindingList.RemoveSort
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private ReadOnly Property SortDirection As ListSortDirection Implements IBindingList.SortDirection
            Get
                Throw New Exception("The method or operation is not implemented.")
            End Get
        End Property

        Private ReadOnly Property SortProperty As PropertyDescriptor Implements IBindingList.SortProperty
            Get
                Throw New Exception("The method or operation is not implemented.")
            End Get
        End Property

        Private ReadOnly Property SupportsChangeNotification As Boolean Implements IBindingList.SupportsChangeNotification
            Get
                Return True
            End Get
        End Property

        Private ReadOnly Property SupportsSearching As Boolean Implements IBindingList.SupportsSearching
            Get
                Return False
            End Get
        End Property

        Private ReadOnly Property SupportsSorting As Boolean Implements IBindingList.SupportsSorting
            Get
                Return False
            End Get
        End Property

#End Region
#Region "IList Members"
        Private Function Add(ByVal value As Object) As Integer Implements IList.Add
            Throw New Exception("The method or operation is not implemented.")
        End Function

        Private Sub Clear() Implements IList.Clear
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private Function Contains(ByVal value As Object) As Boolean Implements IList.Contains
            Throw New Exception("The method or operation is not implemented.")
        End Function

        Private Function IndexOf(ByVal value As Object) As Integer Implements IList.IndexOf
            Throw New Exception("The method or operation is not implemented.")
        End Function

        Private Sub Insert(ByVal index As Integer, ByVal value As Object) Implements IList.Insert
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private ReadOnly Property IsFixedSize As Boolean Implements IList.IsFixedSize
            Get
                Return True
            End Get
        End Property

        Private ReadOnly Property IsReadOnly As Boolean Implements IList.IsReadOnly
            Get
                Return True
            End Get
        End Property

        Private Sub Remove(ByVal value As Object) Implements IList.Remove
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private Sub RemoveAt(ByVal index As Integer) Implements IList.RemoveAt
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private Property Item(ByVal index As Integer) As Object Implements IList.Item
            Get
                Return fData(index)
            End Get

            Set(ByVal value As Object)
                Throw New Exception("The method or operation is not implemented.")
            End Set
        End Property

#End Region
#Region "ICollection Members"
        Private Sub CopyTo(ByVal array As Array, ByVal index As Integer) Implements ICollection.CopyTo
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Private ReadOnly Property Count As Integer Implements ICollection.Count
            Get
                Return fData.Count
            End Get
        End Property

        Private ReadOnly Property IsSynchronized As Boolean Implements ICollection.IsSynchronized
            Get
                Throw New Exception("The method or operation is not implemented.")
            End Get
        End Property

        Private ReadOnly Property SyncRoot As Object Implements ICollection.SyncRoot
            Get
                Throw New Exception("The method or operation is not implemented.")
            End Get
        End Property

#End Region
#Region "IEnumerable Members"
        Private Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
            Return New CustomChartDataSourceEnumerator(fData)
        End Function

#End Region
#Region "ITypedList Members"
        Private Function GetItemProperties(ByVal listAccessors As PropertyDescriptor()) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
            Dim properties As List(Of CustomChartDataSourcePropertyDescriptor) = New List(Of CustomChartDataSourcePropertyDescriptor)()
            properties.Add(New CustomChartDataSourcePropertyDescriptor(CustomChartDataSourceProperty.Series))
            properties.Add(New CustomChartDataSourcePropertyDescriptor(CustomChartDataSourceProperty.Arguments))
            properties.Add(New CustomChartDataSourcePropertyDescriptor(CustomChartDataSourceProperty.Values))
            Return New PropertyDescriptorCollection(properties.ToArray())
        End Function

        Private Function GetListName(ByVal listAccessors As PropertyDescriptor()) As String Implements ITypedList.GetListName
            Throw New Exception("The method or operation is not implemented.")
        End Function

#End Region
#Region "IDisposable Members"
        Public Sub Dispose() Implements IDisposable.Dispose
            If fPivot IsNot Nothing Then
                RemoveHandler fPivot.CellSelectionChanged, AddressOf OnPivotCellSelectionChanged
                fPivot = Nothing
            End If
        End Sub

#End Region
        Public Class CustomChartDataSourceEnumerator
            Implements IEnumerator

            Private fData As List(Of Object())

            Private fPosition As Integer = -1

            Friend Sub New(ByVal data As List(Of Object()))
                fData = data
            End Sub

#Region "IEnumerator Members"
            Private ReadOnly Property Current As Object Implements IEnumerator.Current
                Get
                    If fPosition < 0 OrElse fPosition >= fData.Count Then Return Nothing
                    Return fData(fPosition)
                End Get
            End Property

            Private Function MoveNext() As Boolean Implements IEnumerator.MoveNext
                If fPosition >= fData.Count Then Return False
                fPosition += 1
                Return True
            End Function

            Private Sub Reset() Implements IEnumerator.Reset
                fPosition = -1
            End Sub
#End Region
        End Class

        Public Class CustomChartDataSourcePropertyDescriptor
            Inherits PropertyDescriptor

            Private fPropertyType As CustomChartDataSourceProperty

            Public Sub New(ByVal propertyType As CustomChartDataSourceProperty)
                MyBase.New(propertyType.ToString(), Nothing)
                fPropertyType = propertyType
            End Sub

            Public Overrides Function CanResetValue(ByVal component As Object) As Boolean
                Return False
            End Function

            Public Overrides ReadOnly Property ComponentType As Type
                Get
                    Return GetType(Object())
                End Get
            End Property

            Public Overrides Function GetValue(ByVal component As Object) As Object
                Dim source As Object() = TryCast(component, Object())
                If source Is Nothing OrElse source.Length <> 3 Then Throw New ArgumentException("The array of three values expected")
                Return source(CInt(fPropertyType))
            End Function

            Public Overrides ReadOnly Property IsReadOnly As Boolean
                Get
                    Return True
                End Get
            End Property

            Public Overrides ReadOnly Property PropertyType As Type
                Get
                    Return If(fPropertyType = CustomChartDataSourceProperty.Values, GetType(Decimal), GetType(String))
                End Get
            End Property

            Public Overrides Sub ResetValue(ByVal component As Object)
                Throw New Exception("The method or operation is not implemented.")
            End Sub

            Public Overrides Sub SetValue(ByVal component As Object, ByVal value As Object)
                Throw New Exception("The method or operation is not implemented.")
            End Sub

            Public Overrides Function ShouldSerializeValue(ByVal component As Object) As Boolean
                Throw New Exception("The method or operation is not implemented.")
            End Function
        End Class

        Public Enum CustomChartDataSourceProperty
            Series
            Arguments
            Values
        End Enum
    End Class
End Namespace
